sources/sql_database/schema_types.py

```diff
@@ -20,20 +23,42 @@
     RowAny: TypeAlias = Type[Any]
 
 
+TTypeConversionCallback = Callable[
+    [sqltypes.TypeEngine[Any]], Optional[Union[sqltypes.TypeEngine[Any], Type[sqltypes.TypeEngine[Any]]]]
+]
+
+
 def sqla_col_to_column_schema(
-    sql_col: ColumnAny, add_precision: bool = False
+    sql_col: ColumnAny,
+    reflection_level: ReflectionLevel,
+    type_conversion_callback: Optional[TTypeConversionCallback] = None,
 ) -> Optional[TColumnSchema]:
     """Infer dlt schema column type from an sqlalchemy type.
 
     If `add_precision` is set, precision and scale is inferred from that types that support it,
     such as numeric, varchar, int, bigint. Numeric (decimal) types have always precision added.
     """
-    sql_t = sql_col.type
     col: TColumnSchema = {
         "name": sql_col.name,
-        "data_type": None,  # set that later
         "nullable": sql_col.nullable,
     }
+    if reflection_level == "minimal":
+        return col
+
+    sql_t = sql_col.type
+
+    sql_t = type_conversion_callback(sql_t) if type_conversion_callback else sql_t
+    if type_conversion_callback:
+        sql_t = type_conversion_callback(sql_t)
+        # Check if sqla type class rather than instance is returned
+        if sql_t is not None and isinstance(sql_t, type):
+            sql_t = sql_t()
+
+    if sql_t is None:
+        # Column ignored by callback
+        return None
```

Comment:
please take a look #488 - this is actually a bug that will prevent pyarrow backend from working.
we are converting a list of rows into ndarray and then transpose and split it. to do that a list of columns is used - so we may still return a column but without data type.

please take a look and write a test ie. using GEOMETRY or some other postgres type that we do not handle